src/Associations.cc

```diff
@@ -265,61 +261,88 @@ void Associations::associateRefStars(double matchCutInArcSec, const Gtransfo *gt
-    */
+
+    // first pass: remove objects that have less than a certain number of measurements.
     for (auto const &ccdImage : ccdImageList) {
         MeasuredStarList &catalog = ccdImage->getCatalogForFit();
+        // Iteration happens internal to the loop, as we may delete measuredStars from catalog.
         for (MeasuredStarIterator mi = catalog.begin(); mi != catalog.end();) {
             MeasuredStar &mstar = **mi;
 
-            auto fstar = mstar.getFittedStar();
-            if (!fstar) {
+            auto fittedStar = mstar.getFittedStar();
+            // measuredStar has no fittedStar: move on.
+            if (fittedStar == nullptr) {
                 ++mi;
                 continue;
             }
 
-            /*  keep FittedStar's which either have a minimum number of
-                measurements, or are matched to a RefStar
-            */
-            if (!fstar->getRefStar() && fstar->getMeasurementCount() < minMeasurements) {
-                auto f = std::const_pointer_cast<FittedStar>(fstar);
-                f->getMeasurementCount()--;
-                mi = catalog.erase(mi);
-            } else
+            // keep FittedStars which either have a minimum number of
+            // measurements, or are matched to a RefStar
+            if (!fittedStar->getRefStar() && fittedStar->getMeasurementCount() < minMeasurements) {
+                fittedStar->getMeasurementCount()--;
+                mi = catalog.erase(mi);  // mi now points to the next measuredStar.
+            } else {
                 ++mi;
+            }
         }  // end loop on objects in catalog
     }      // end loop on catalogs
 
-    /* now FittedStars with less than minMeasurements should have
-       zero measurementCount; */
-
+    // now FittedStars with less than minMeasurements should have zero measurementCount.
     for (FittedStarIterator fi = fittedStarList.begin(); fi != fittedStarList.end();) {
-        if ((*fi)->getMeasurementCount() == 0)
+        if ((*fi)->getMeasurementCount() == 0) {
             fi = fittedStarList.erase(fi);
-        else
+        } else {
             ++fi;
+        }
     }
 
     LOGLS_INFO(_log, "Fitted stars after measurement # cut: " << fittedStarList.size());
 }
 
+void Associations::normalizeFittedStars() const {
+    // Clear positions in order to take the average of the measuredStars.
+    for (auto &fittedStar : fittedStarList) {
+        fittedStar->x = 0.0;
+        fittedStar->y = 0.0;
+        fittedStar->setFlux(0.0);
+    }
+
+    // Iterate over measuredStars to add their values into their fittedStars
+    for (auto const &ccdImage : ccdImageList) {
+        const Gtransfo *toCommonTangentPlane = ccdImage->getPix2CommonTangentPlane();
+        MeasuredStarList &catalog = ccdImage->getCatalogForFit();
+        for (auto &mi : catalog) {
+            auto fittedStar = mi->getFittedStar();
+            assert(fittedStar != nullptr);  // This must be true, if selectFittedStars() was called.
```

Comment:
I'd prefer an exception to `assert`. Exiting a program from low-level code can cause a lot of trouble from stuff not getting cleaned up (e.g., open files, incomplete logs, any on-failure guarantees offered by client code). Basically, there's too much room for unintended consequences.

Reply:
I'm still not sure about when to choose `assert` vs. `throw` in C++. I prefer `throw` here too, especially since it can include an explanatory message.

Reply:
You should `assert` only for something that cannot happen, where the computer is so corrupted that proceeding any further would be useless.  `throw` can potentially be caught and lead to continued processing. `assert` should not, however, be used merely to state a precondition for some code.

For example, if you knew that `mi->getFittedStar()` could never return `nullptr`, this `assert` would be fine.  But if it's just passing along something that came from elsewhere, it should be a `throw`.